### PR TITLE
Cleanup win32defines

### PR DESCRIPTION
**Pull request by zhekov**
Lots of words coming - sit tight, or skip to TL;DR aka p.5 below.

---

\1. About WINVER, _WIN32_WINNT and _WIN32_IE

WINVER/_WIN32_WINNT define a minimum Windows API level. There are two of them, because at the times of Windows 3.x/WinNT 3.x and Windows 9x/WinNT 4.x a single WINVER would mean targeting both Windows API X and NT API Y, which are two very different things.

_WIN32_IE exists because at least 9x was sometines updated by installing a new IE, which replaced core libraries. So some symbols depend on _WIN32_IE being certain version. Lame.

The above created some confusion. After 0x0500 (Win2K), the things should have settled, because Windows 9x ended at 0x04XX - but in practice, sometimes WINVER >= 0x0500 is checked. In the newer MinGW-s, such checks are almost extinct, and API levels < 0x0500 are not properly supported anyway - some headers will break if WINVER/_WIN32_WINNT is followed strictly.

\2. About VC_EXTRALEAN and WIN32_LEAN_AND_MEAN.

These exclude parts of the Windows API to speed up compilation. VC_EXTRALEAN is normally for MFC, and WIN32_LEAN_AND_MEAN for everything else (us), though it seems that both can be used almost without problems.

In my tests, with both defined, the "speedup" was inside the margin of error under Waf. Actually, once I hot a bit higher time with them defined, probably due to some other process taking a bit of time. Considering the compilation Scintilla and glib/gtk+, defining these has zero effect.

Such exclusions should be defined per-module. By defining them globally, we block access to the excluded API-s for all modules by default (unless specific headers are included), without any good reason. And we possibly create problems for whoever may decide to use STRICT (a popular -D which defines the Windows API more strictly, thus enabling more compiler checks for WinAPI calls).

Also note that WIN32_LEAN_AND_MEAN excludes Windows Sockets among the rest. But we get away with that, be it because socket.c includes winsock2 directly, or for some other reason...

\3. Polular misconceptions.

"WINVER/_WIN32_WINNT must be defined to use windows.h / WinAPI properly"

Only if you need symbols not included by your default API - higher level, or specific to _WIN32_IE, or something else.

"WINVER/_WIN32_WINNT is defined if you include windows.h"

With the newer MinGW w32api-s, including any standard header (tested with stdio and stdlib) results in including lower-level system headers that in turn define WINVER/_WIN32_WINNT. Absent an explicit #define, these are defined to the default version for this set of WinAPI headers.

"WINVER/_WIN32_WINNT must be the same for all modules"

Only if you pass API-level dependent Windows structures between the modules. We don't pass any Windows structures, and our WINVER/_WIN32_WINNT are different for socket/spawn/win32 and everything else - see the previous paragraph.

"defining WINVER/_WIN32_WINNT 0x0501 will make your program XP only"

Using a function enabled by 0x0501 will do that, not the define by itself. Running the program on < XP will display something like "DLL X lacks function Y". We can target a specific windows version by using the linker, and the message will change to something like "The specified program requires a newer version of Windows".

\4. Compiling Geany with MinGW.

scintilla-3.5.6 compiles with MinGW w32api-3.17-2, which defaults to API 0x0400. *All* earlier MinGW w32api-s are 0x0400 as well. Levels 3.x are ancient, for 16-bit Windows and the long forgotten WinNT 3.x, and we are not interested by them.

So we need to find whether we need something outside the default API level 0x0400. Having such a MinGW, I simply tried. Turns out, we need SHGetFolderPathAndSubDirW and SHGFP_TYPE only, for win32.c only. As properly documented in win32defines, that needs API 0x0501 for the former, and _WIN32_IE 0x0500 for the latter. sockets/spawn require *nothing*.

In both MinGW and MSVC, SHGetFolderPathAndSubDirW is dependent (directly or indirectly) on _WIN32_WINNT, and not WINVER.

---

TL;DR

\5. How to make the things "right"?

Minimal: move _WIN32_WINNT 0x0501 and _WIN32_IE 0x0500 to win32.c and remove win32defines. Yes, that actually works, with the oldest API level 0x0400. Advantages: "Define what you use" for win32, and don't define what you don't use/need - especially LEAN_AND_MEAN for socket and windows-only macros for socket/spawn under *nix. No, the last one will not make any problems, it's just ugly. Disadvantages: none (see p.4 above).

Maximal: define both WINVER and _WIN32_WINNT in a header that is included by all modules under Windows with all build systems under Win~1, and set the linker target to Windows XP. Advantages: identical API levels for all modules, a bit better message for the users when attempting to run Geany under Win~1 < XP. Disadvantages: requires work on the build systems.

Opportunistic: leave everything as-is - what the hell, even paid workers wouldn't bother. :)

This PR implements the minimal solution.

### CODE REVIEW

**Comment by eht16** [comment]
Wow, great explanation. Many thanks for this.

I'm happy with the WINVER/_WIN32_WINNT change.

Not so sure about WIN32_LEAN_AND_MEAN: it seems I introduced it in b54eb0a180cf288c41087abd655bbe567e8edca1. And I remember there was a reason for, certainly it was not to speed up compilation but rather fix compilation or some behaviour. Unfortunately, I don't remember the exact reason and I also didn't document it. Shame on me.
I just compiled Geany with your patch and it seems to work without WIN32_LEAN_AND_MEAN now.
No compiler warnings/errors, no obvious runtime problems.

So, since I don't know why I added WIN32_LEAN_AND_MEAN back in 2007, it seems not necessary anymore today. Be it because of newer Mingw versions or changed code in Geany or whatever.

I have no problems using this, as Geany compiles and runs with these changes (for me) and the changes sound reasonable to me, according to the explanation above.

**Comment by zhekov** [comment]
About WIN32_LEAN_AND_MEAN, it may be that parts of the WinAPI did not compile. Microsoft never followed the standards too strictly (WIN16 violated it outright), and their compiler is very lax ("static enum" :) anyone?). If that was the case, probably MS or the MinGW guys fixed the headers later.

Another possibility is that something MS/Windows specific was unsupported or buggy in gcc at the time, for example ms_struct / mms-bitfields... Anyway, since our scintilla *requires* gcc-4.7.1+, and the respective winapi >= 3.17 is confirmed to work, we should be fine now.

**Comment by elextr** [comment]
Just beware there is one version of gcc (4.8.1?) where there was a regression in its matching the sizes of things on Windows, it may not work.

**Comment by eht16** [comment]
@elextr can we do anything about a broken gcc version out there? And is this related to the suggested changes here?

The rest LGTM.

**Comment by elextr** [comment]
My Googleing finally found it, it is gcc 4.8.1 in mingw C++ floating point size mismatch for long double.

So it probably doesn't affect us, since I don't think Scintilla passes a long double to system libraries when we use GTK.

So I wouldn't worry about it any more.

**Comment by b4n** [comment]
I didn't actually test, but LGTM.

**Comment by zhekov** [comment]
@eht16 

> can we do anything about a broken gcc version out there? And is this related to the suggested changes here?

To the second question: it's not related. An API declaration will be the same no matter whether you include the default windows.h or the LEAN_AND_MEAN-.h plus extra headers, and enable it with a global or local _WIN32_WINNT/WINVER.

**Comment by eht16** [comment]
Thanks for clarifying.
